### ISSUE TITLE
toggle problems with rotated monitor

### ISSUE THREAD

**Issue by jcpearson** [tb2097/wacom-gui#2]
I'm having problems getting the '--toggle' option to work when one monitor is rotated by 90 degrees

I'm using CentOS 7.4 with Mate and two 1920x1200 monitors using the Nvidia driver

The right-hand monitor is rotated 90 degrees left (via nvidia-settings) i.e. in portrait mode

Using wacom-gui in GUI mode, I can set the Wacom tablet to use left, right or both monitors - which works fine - however when using in 'toggle' mode, it crashes with:
Traceback (most recent call last):
  File "/usr/local/wacom-gui/wacom-gui.py", line 338, in <module>
    main()    
  File "/usr/local/wacom-gui/wacom-gui.py", line 178, in main
    parseArgs(sys.argv)
  File "/usr/local/wacom-gui/wacom-gui.py", line 193, in parseArgs
    toggleScreens()
  File "/usr/local/wacom-gui/wacom-gui.py", line 211, in toggleScreens
    coords = getTabletArea(mod[0])
  File "/usr/local/wacom-gui/wacom-gui.py", line 286, in getTabletArea
    display[0][0] = float(screens[1][2]) / float(totalResolution[0])
IndexError: list index out of range

The problem is that xrandr returns a slightly different result when monitors are rotated - in my case:

    DVI-I-1 connected 1920x1200+0+360 (normal left inverted right x axis y axis) 519mm x 324mm
    DP-1 connected 1200x1920+1920+0 left (normal left inverted right x axis y axis) 518mm x 324mm

i.e. the extra 'left' in the output for the rotated monitor (which I guess mean the monitor is rotated to the left ...)

I can 'fix' this by replacing the 'get screen dimensions/coords' code in wacom-gui.py with:

    info = os.popen("xrandr | grep ' connected'").read().strip().split('\n')
    screens = []
    for screen in info:
        scr = re.findall(r'\d+x\d+\+\d+\+\d+', screen)[0]
        scr = scr.replace('x','+')
        scr = scr.split('+')
        scr.insert(3,scr.pop(0))
        scr.insert(3,scr.pop(0))
        if screen.find('primary') == -1:
            screens.append(scr)
        else:
            screens.insert(0, scr)

However, it now only allows me to toggle between full screen and the left monitor ... it won't ever toggle to the right monitor

In fact, I can only get it to toggle between left, right and both monitors when both monitors are not rotated and aligned with the same top-left 'y' coordinate - if there is any offset in the vertical , then I can not toggle to the right monitor

I haven't had a chance to workout how the code is selecting which monitor to toggle to, but I suspect it is not taking into account any offsets between the monitors ?

Thanks

**Comment by tb2097** [comment]
I think I might've fixed this in another section of the code but not with the --toggle option...
If you open up the GUI and try to toggle between the options there does it correctly detect left/right/full or does it crash as well?  I *think* it should work correctly via the GUI, which if it does means I just need to do some tweaking of the code.  I won't have time to invest is checking this until next week though.  If you can test/update your response it'll help me narrow down (hopefully) where the fix is needed.

**Comment by jcpearson** [comment]
Works fine in the GUI - problem occurs just with the --toggle option

**Comment by jcpearson** [comment]
It looks like it is possible to achieve the same as the '--toggle' option by using:

    xsetwacom --set "$tabletName" MapToOutput next

which cycles through mapping the tablet to all screens and then whole desktop

**Comment by jcpearson** [comment]
I believe the problem is a rounding issue - the following patch allows '--toggle' to work as expected:

```
--- ./wacom-gui/wacom-gui.py.dist       2018-02-02 18:10:47.000000000 +0000
+++ ./wacom-gui/wacom-gui.py    2018-04-05 15:43:52.044730470 +0100
@@ -278,7 +278,7 @@ def getTabletArea(*args):
     isLeft = True
     for i in range(0, 3):
         for j in range(0, 3):
-            if float(tabletScreenCoords[(i, j)]) != float(display[i][j]):
+            if round(float(tabletScreenCoords[(i, j)]),6) != round(float(display[i][j]),6):
                 isLeft = False
                 break
     if isLeft:
```

**Comment by tb2097** [comment]
Tested/implemented patch, part of release 0.1.9.  Thanks for finding this!

**Comment by jcpearson** [comment]
Thanks - but the original problem hasn't been fixed (problem with the extra 'left' in xrandr output when a screen is rotated to the left) - the following patch should fix this:

```
--- ./wacom-gui/wacom-gui.py.dist       2018-02-02 18:10:47.000000000 +0000
+++ ./wacom-gui/wacom-gui.py    2018-04-09 11:53:40.284044530 +0100
@@ -246,15 +246,15 @@ def getTabletArea(*args):
     info = os.popen("xrandr | grep ' connected'").read().strip().split('\n')
     screens = []
     for screen in info:
-        screen = screen.split('(')[0].strip().rsplit(' ', 1)
-        screen[1] = screen[1].replace('x','+')
-        screen[1] = screen[1].split('+')
-        screen[1].insert(3,screen[1].pop(0))
-        screen[1].insert(3,screen[1].pop(0))
-        if screen[0].find('primary') == -1:
-            screens.append(screen[1])
+        scr = re.findall(r'\d+x\d+\+\d+\+\d+', screen)[0]
+        scr = scr.replace('x', '+')
+        scr = scr.split('+')
+        scr.insert(3, scr.pop(0))
+        scr.insert(3, scr.pop(0))
+        if screen.find('primary') == -1:
+            screens.append(scr)
         else:
-            screens.insert(0, screen[1])
+            screens.insert(0, scr)
     totalResolution = os.popen("xdpyinfo | grep dimensions | awk '{print $2}' | awk -Fx '{print $1, $2}'").read()
     totalResolution = totalResolution.split()

```
